File: sigma-tree/src/chain/ergo_box.rs
//! Ergo box

pub mod box_id;
pub mod box_value;
pub mod register;

#[cfg(feature = "with-serde")]
use super::json;
use super::{
    digest32::blake2b256_hash,
    token::{TokenAmount, TokenId},
    transaction::TxId,
};
use crate::{
    ergo_tree::ErgoTree,
    serialization::{
        ergo_box::{parse_box_with_indexed_digests, serialize_box_with_indexed_digests},
        sigma_byte_reader::SigmaByteRead,
        sigma_byte_writer::SigmaByteWrite,
        SerializationError, SigmaSerializable,
    },
};
use box_id::BoxId;
use box_value::BoxValue;
use indexmap::IndexSet;
use register::NonMandatoryRegisters;
#[cfg(feature = "with-serde")]
use serde::{Deserialize, Serialize};
#[cfg(feature = "with-serde")]
use std::convert::TryFrom;
use std::io;
#[cfg(feature = "with-serde")]
use thiserror::Error;

/// Box (aka coin, or an unspent output) is a basic concept of a UTXO-based cryptocurrency.
/// In Bitcoin, such an object is associated with some monetary value (arbitrary,
/// but with predefined precision, so we use integer arithmetic to work with the value),
/// and also a guarding script (aka proposition) to protect the box from unauthorized opening.
///
/// In other way, a box is a state element locked by some proposition (ErgoTree).
///
/// In Ergo, box is just a collection of registers, some with mandatory types and semantics,
/// others could be used by applications in any way.
/// We add additional fields in addition to amount and proposition~(which stored in the registers R0 and R1).
/// Namely, register R2 contains additional tokens (a sequence of pairs (token identifier, value)).
/// Register R3 contains height specified by user (protocol checks if it was <= current height when
/// transaction was accepted) and also transaction identifier and box index in the transaction outputs.
/// Registers R4-R9 are free for arbitrary usage.
///
/// A transaction is unsealing a box. As a box can not be open twice, any further valid transaction
/// can not be linked to the same box.
#[cfg_attr(feature = "with-serde", derive(Serialize, Deserialize))]
#[cfg_attr(
    feature = "with-serde",
    serde(try_from = "json::ergo_box::ErgoBoxFromJson")
)]
#[derive(PartialEq, Eq, Debug, Clone)]
pub struct ErgoBox {
    #[cfg_attr(feature = "with-serde", serde(rename = "boxId"))]
    box_id: BoxId,
    /// amount of money associated with the box
    #[cfg_attr(feature = "with-serde", serde(rename = "value"))]
    pub value: BoxValue,
    /// guarding script, which should be evaluated to true in order to open this box
    #[cfg_attr(
        feature = "with-serde",
        serde(rename = "ergoTree", with = "json::ergo_tree")
    )]
    pub ergo_tree: ErgoTree,
    /// secondary tokens the box contains
    #[cfg_attr(feature = "with-serde", serde(rename = "assets"))]
    pub tokens: Vec<TokenAmount>,
    ///  additional registers the box can carry over
    #[cfg_attr(feature = "with-serde", serde(rename = "additionalRegisters"))]
    pub additional_registers: NonMandatoryRegisters,
    /// height when a transaction containing the box was created.
    /// This height is declared by user and should not exceed height of the block,
    /// containing the transaction with this box.
    #[cfg_attr(feature = "with-serde", serde(rename = "creationHeight"))]
    pub creation_height: u32,
    /// id of transaction which created the box
    #[cfg_attr(feature = "with-serde", serde(rename = "transactionId"))]
    pub transaction_id: TxId,
    /// number of box (from 0 to total number of boxes the transaction with transactionId created - 1)
    #[cfg_attr(feature = "with-serde", serde(rename = "index"))]
    pub index: u16,
}

impl ErgoBox {
    /// Crate new box
    pub fn new(
        value: BoxValue,
        ergo_tree: ErgoTree,
        tokens: Vec<TokenAmount>,
        additional_registers: NonMandatoryRegisters,
        creation_height: u32,
        transaction_id: TxId,
        index: u16,
    ) -> ErgoBox {
        let box_with_zero_id = ErgoBox {
            box_id: BoxId::zero(),
            value,
            ergo_tree,
            tokens,
            additional_registers,
            creation_height,
            transaction_id,
            index,
        };
        let box_id = box_with_zero_id.calc_box_id();
        ErgoBox {
            box_id,
            ..box_with_zero_id
        }
    }

    /// Box id (Blake2b256 hash of serialized box)
    pub fn box_id(&self) -> BoxId {
        self.box_id.clone()
    }

    /// Create ErgoBox from ErgoBoxCandidate by adding transaction id
    /// and index of the box in the transaction
    pub fn from_box_candidate(
        box_candidate: &ErgoBoxCandidate,
        transaction_id: TxId,
        index: u16,
    ) -> ErgoBox {
        let box_with_zero_id = ErgoBox {
            box_id: BoxId::zero(),
            value: box_candidate.value,
            ergo_tree: box_candidate.ergo_tree.clone(),
            tokens: box_candidate.tokens.clone(),
            additional_registers: box_candidate.additional_registers.clone(),
            creation_height: box_candidate.creation_height,
            transaction_id,
            index,
        };
        let box_id = box_with_zero_id.calc_box_id();
        ErgoBox {
            box_id,
            ..box_with_zero_id
        }
    }

    fn calc_box_id(&self) -> BoxId {
        let bytes = self.sigma_serialise_bytes();
        BoxId(blake2b256_hash(&bytes))
    }
}

/// Assets that ErgoBox holds
pub trait ErgoBoxAssets {
    /// Box value
    fn value(&self) -> BoxValue;
    /// Tokens (ids and amounts)
    fn tokens(&self) -> Vec<TokenAmount>;
}

/// Simple struct to hold ErgoBoxAssets values
pub struct ErgoBoxAssetsData {
    /// Box value
    pub value: BoxValue,
    /// Tokens
    pub tokens: Vec<TokenAmount>,
}

impl ErgoBoxAssets for ErgoBoxAssetsData {
    fn value(&self) -> BoxValue {
        self.value
    }

    fn tokens(&self) -> Vec<TokenAmount> {
        self.tokens.clone()
    }
}

impl ErgoBoxAssets for ErgoBoxCandidate {
    fn value(&self) -> BoxValue {
        self.value
    }

    fn tokens(&self) -> Vec<TokenAmount> {
        self.tokens.clone()
    }
}

impl ErgoBoxAssets for ErgoBox {
    fn value(&self) -> BoxValue {
        self.value
    }

    fn tokens(&self) -> Vec<TokenAmount> {
        self.tokens.clone()
    }
}

/// id of the ergo box
pub trait ErgoBoxId {
    /// Id of the ergo box
    fn box_id(&self) -> BoxId;
}

impl ErgoBoxId for ErgoBox {
    fn box_id(&self) -> BoxId {
        self.box_id.clone()
    }
}

/// Errors on parsing ErgoBox from JSON
#[cfg(feature = "with-serde")]
#[derive(Error, PartialEq, Eq, Debug, Clone)]
pub enum ErgoBoxFromJsonError {
    /// Box id parsed from JSON differs from calculated from box serialized bytes
    #[error("Box id parsed from JSON differs from calculated from box serialized bytes")]
    InvalidBoxId,
}

#[cfg(feature = "with-serde")]
impl TryFrom<json::ergo_box::ErgoBoxFromJson> for ErgoBox {
    type Error = ErgoBoxFromJsonError;
    fn try_from(box_json: json::ergo_box::ErgoBoxFromJson) -> Result<Self, Self::Error> {
        let box_with_zero_id = ErgoBox {
            box_id: BoxId::zero(),
            value: box_json.value,
            ergo_tree: box_json.ergo_tree,
            tokens: box_json.tokens,
            additional_registers: box_json.additional_registers,
            creation_height: box_json.creation_height,
            transaction_id: box_json.transaction_id,
            index: box_json.index,
        };
        let box_id = box_with_zero_id.calc_box_id();
        let ergo_box = ErgoBox {
            box_id,
            ..box_with_zero_id
        };
        if ergo_box.box_id() == box_json.box_id {
            Ok(ergo_box)
        } else {
            Err(ErgoBoxFromJsonError::InvalidBoxId)
        }
    }
}

impl SigmaSerializable for ErgoBox {
    fn sigma_serialize<W: SigmaByteWrite>(&self, w: &mut W) -> Result<(), io::Error> {
        let ergo_tree_bytes = self.ergo_tree.sigma_serialise_bytes();
        serialize_box_with_indexed_digests(
            &self.value,
            ergo_tree_bytes,
            &self.tokens,
            &self.additional_registers,
            self.creation_height,
            None,
            w,
        )?;
        self.transaction_id.sigma_serialize(w)?;
        w.put_u16(self.index)?;
        Ok(())
    }
    fn sigma_parse<R: SigmaByteRead>(r: &mut R) -> Result<Self, SerializationError> {
        let box_candidate = ErgoBoxCandidate::parse_body_with_indexed_digests(None, r)?;
        let tx_id = TxId::sigma_parse(r)?;
        let index = r.get_u16()?;
        Ok(ErgoBox::from_box_candidate(&box_candidate, tx_id, index))
    }
}

/// Contains the same fields as `ErgoBox`, except if transaction id and index,
/// that will be calculated after full transaction formation.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ErgoBoxCandidate {
    /// amount of money associated with the box
    pub value: BoxValue,
    /// guarding script, which should be evaluated to true in order to open this box
    pub ergo_tree: ErgoTree,
    /// secondary tokens the box contains
    pub tokens: Vec<TokenAmount>,
    ///  additional registers the box can carry over
    pub additional_registers: NonMandatoryRegisters,
    /// height when a transaction containing the box was created.
    /// This height is declared by user and should not exceed height of the block,
    /// containing the transaction with this box.
    pub creation_height: u32,
}

impl ErgoBoxCandidate {
    /// create box with value guarded by ErgoTree
    pub fn new(value: BoxValue, ergo_tree: ErgoTree, creation_height: u32) -> ErgoBoxCandidate {
        // TODO: check for min BoxValue and return error
        ErgoBoxCandidate {
            value,
            ergo_tree,
            tokens: vec![],
            additional_registers: NonMandatoryRegisters::empty(),
            creation_height,
        }
    }

    /// Box serialization with token ids optionally saved in transaction
    /// (in this case only token index is saved)
    pub fn serialize_body_with_indexed_digests<W: SigmaByteWrite>(
        &self,
        token_ids_in_tx: Option<&IndexSet<TokenId>>,
        w: &mut W,
    ) -> Result<(), io::Error> {
        serialize_box_with_indexed_digests(
            &self.value,
            self.ergo_tree.sigma_serialise_bytes(),
            &self.tokens,
            &self.additional_registers,
            self.creation_height,
            token_ids_in_tx,
            w,
        )
    }

    /// Box deserialization with token ids optionally parsed in transaction
    pub fn parse_body_with_indexed_digests<R: SigmaByteRead>(
        digests_in_tx: Option<&IndexSet<TokenId>>,
        r: &mut R,
    ) -> Result<ErgoBoxCandidate, SerializationError> {
        parse_box_with_indexed_digests(digests_in_tx, r)
    }
}

impl SigmaSerializable for ErgoBoxCandidate {
    fn sigma_serialize<W: SigmaByteWrite>(&self, w: &mut W) -> Result<(), io::Error> {
        self.serialize_body_with_indexed_digests(None, w)
    }
    fn sigma_parse<R: SigmaByteRead>(r: &mut R) -> Result<Self, SerializationError> {
        ErgoBoxCandidate::parse_body_with_indexed_digests(None, r)
    }
}

impl From<ErgoBox> for ErgoBoxCandidate {
    fn from(b: ErgoBox) -> Self {
        ErgoBoxCandidate {
            value: b.value,
            ergo_tree: b.ergo_tree,
            tokens: b.tokens,
            additional_registers: b.additional_registers,
            creation_height: b.creation_height,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::serialization::sigma_serialize_roundtrip;
    use proptest::{arbitrary::Arbitrary, collection::vec, prelude::*};

    impl Arbitrary for ErgoBoxCandidate {
        type Parameters = super::box_value::tests::ArbBoxValueRange;

        fn arbitrary_with(args: Self::Parameters) -> Self::Strategy {
            (
                any_with::<BoxValue>(args),
                any::<ErgoTree>(),
                vec(any::<TokenAmount>(), 0..10),
                any::<u32>(),
                any::<NonMandatoryRegisters>(),
            )
                .prop_map(
                    |(value, ergo_tree, tokens, creation_height, additional_registers)| Self {
                        value,
                        ergo_tree,
                        tokens,
                        additional_registers,
                        creation_height,
                    },
                )
                .boxed()
        }
        type Strategy = BoxedStrategy<Self>;
    }

    impl Arbitrary for ErgoBox {
        type Parameters = super::box_value::tests::ArbBoxValueRange;

        fn arbitrary_with(args: Self::Parameters) -> Self::Strategy {
            (
                any_with::<ErgoBoxCandidate>(args),
                any::<TxId>(),
                any::<u16>(),
            )
                .prop_map(|(box_candidate, tx_id, index)| {
                    Self::from_box_candidate(&box_candidate, tx_id, index)
                })
                .boxed()
        }
        type Strategy = BoxedStrategy<Self>;
    }

    proptest! {

        #[test]
        fn ergo_box_candidate_ser_roundtrip(v in any::<ErgoBoxCandidate>()) {
            prop_assert_eq![sigma_serialize_roundtrip(&v), v];
        }

        #[test]
        fn ergo_box_ser_roundtrip(v in any::<ErgoBox>()) {
            prop_assert_eq![sigma_serialize_roundtrip(&v), v];
        }
    }
}

File: bindings/ergo-lib-ios/Sources/ErgoLib/ErgoBox.swift
/// Box (aka coin, or an unspent output) is a basic concept of a UTXO-based cryptocurrency.
/// In Bitcoin, such an object is associated with some monetary value (arbitrary,
/// but with predefined precision, so we use integer arithmetic to work with the value),
/// and also a guarding script (aka proposition) to protect the box from unauthorized opening.
///
/// In other way, a box is a state element locked by some proposition (ErgoTree).
///
/// In Ergo, box is just a collection of registers, some with mandatory types and semantics,
/// others could be used by applications in any way.
/// We add additional fields in addition to amount and proposition~(which stored in the registers R0 and R1).
/// Namely, register R2 contains additional tokens (a sequence of pairs (token identifier, value)).
/// Register R3 contains height when block got included into the blockchain and also transaction
/// identifier and box index in the transaction outputs.
/// Registers R4-R9 are free for arbitrary usage.
///
/// A transaction is unsealing a box. As a box can not be open twice, any further valid transaction
//! can not be linked to the same box.
import Foundation
import ErgoLibC

/// Box id (32-byte digest)
class BoxId {
    internal var pointer: BoxIdPtr
    
    /// Parse box id (32 byte digest) from base16-encoded string
    init(withString str: String) throws {
        var ptr: BoxIdPtr?
        let error = str.withCString { cs in
            ergo_wallet_box_id_from_str(cs, &ptr)
        }
        try checkError(error)
        self.pointer = ptr!
    }
    
    /// Takes ownership of an existing ``BoxIdPtr``. Note: we must ensure that no other instance
    /// of ``BoxId`` can hold this pointer.
    internal init(withRawPointer ptr: BoxIdPtr) {
        self.pointer = ptr
    }
    
    /// Returns byte array (32 bytes) representation
    func toBytes() -> [UInt8] {
        var bytes = Array.init(repeating: UInt8(0), count: 32)
        ergo_wallet_box_id_to_bytes(self.pointer, &bytes)
        return bytes
    }
     
    /// Base16 encoded string
    func toString() -> String {
        var cStr: UnsafePointer<CChar>?
        ergo_wallet_box_id_to_str(self.pointer, &cStr)
        let str = String(cString: cStr!)
        ergo_wallet_delete_string(UnsafeMutablePointer(mutating: cStr))
        return str
    }
    
    deinit {
        ergo_wallet_box_id_delete(self.pointer)
    }
}

extension BoxId: Equatable {
    static func ==(lhs: BoxId, rhs: BoxId) -> Bool {
        ergo_wallet_box_id_eq(lhs.pointer, rhs.pointer)
    }
}

/// Box value in nanoERGs with bound checks
class BoxValue {
    internal var pointer: BoxValuePtr
    
    /// Create from ``Int64`` with bounds check
    init(fromInt64 : Int64) throws {
        var ptr: BoxValuePtr?
        let error = ergo_wallet_box_value_from_i64(fromInt64, &ptr)
        try checkError(error)
        self.pointer = ptr!
    }
    
    init(withRawPointer ptr: BoxValuePtr) {
        self.pointer = ptr
    }
    
    /// Recommended (safe) minimal box value to use in case box size estimation is unavailable.
    /// Allows box size upto 2777 bytes with current min box value per byte of 360 nanoERGs
    static func SAFE_USER_MIN() -> BoxValue {
        var ptr: BoxValuePtr?
        ergo_wallet_box_value_safe_user_min(&ptr)
        return BoxValue(withRawPointer: ptr!)
    }
    
    /// Number of units inside one ERGO (i.e. one ERG using nano ERG representation)
    static func UNITS_PER_ERGO() -> Int64 {
        return ergo_wallet_box_value_units_per_ergo()
    }
    
    /// Create a new box value which is the sum of the arguments, throwing error if value is out of
    /// bounds. Note that swift forbids integer overflow/underflow by default with hard crash, which
    /// is not desirable here.
    static func sumOf(boxValue0: BoxValue, boxValue1: BoxValue) throws -> BoxValue {
        var ptr: BoxValuePtr?
        let error = ergo_wallet_box_value_sum_of(boxValue0.pointer, boxValue1.pointer, &ptr)
        try checkError(error)
        return BoxValue(withRawPointer: ptr!)
    }
    
    /// Get value as ``Int64``
    func toInt64() -> Int64 {
        return ergo_wallet_box_value_as_i64(self.pointer)
    }
    
    deinit {
        ergo_wallet_box_value_delete(self.pointer)
    }
}

extension BoxValue: Equatable {
    static func ==(lhs: BoxValue, rhs: BoxValue) -> Bool {
        ergo_wallet_box_value_eq(lhs.pointer, rhs.pointer)
    }
}

/// Contains the same fields as ``ErgoBox``, except for transaction id and index, that will be
/// calculated after full transaction formation.  Use ``ErgoBoxCandidateBuilder`` to create an
/// instance.
class ErgoBoxCandidate {
    internal var pointer: ErgoBoxCandidatePtr

    /// Takes ownership of an existing ``ErgoBoxCandidatePtr``. Note: we must ensure that no other
    /// instance of ``ErgoBoxCandidate`` can hold this pointer.
    internal init(withRawPointer pointer: ErgoBoxCandidatePtr) {
        self.pointer = pointer
    }
    
    /// Returns value (ErgoTree constant) stored in the register or `nil` if the register is empty
    func getRegisterValue(registerId: NonMandatoryRegisterId) -> Constant? {
        var constantPtr: ConstantPtr?
        let res = ergo_wallet_ergo_box_candidate_register_value(self.pointer, registerId.rawValue, &constantPtr)
        assert(res.error == nil)
        if res.is_some {
            return Constant(withRawPointer: constantPtr!)
        } else {
            return nil
        }
    }
    
    /// Get box creation height
    func getCreationHeight() -> UInt32 {
        return ergo_wallet_ergo_box_candidate_creation_height(self.pointer)
    }

    /// Get tokens for box
    func getTokens() -> Tokens {
        var tokensPtr: TokensPtr?
        ergo_wallet_ergo_box_candidate_tokens(self.pointer, &tokensPtr)
        return Tokens(withPtr: tokensPtr!)
    }
    
    /// Get ergo tree for box
    func getErgoTree() -> ErgoTree {
        var ergoTreePtr: ErgoTreePtr?
        ergo_wallet_ergo_box_candidate_ergo_tree(self.pointer, &ergoTreePtr)
        return ErgoTree(withRawPointer: ergoTreePtr!)
    }
    
    /// Get box value
    func getBoxValue() -> BoxValue {
        var boxValuePtr: BoxValuePtr?
        ergo_wallet_ergo_box_candidate_box_value(self.pointer, &boxValuePtr)
        return BoxValue(withRawPointer: boxValuePtr!)
    }
    
    deinit {
        ergo_wallet_ergo_box_candidate_delete(self.pointer)
    }
}

extension ErgoBoxCandidate: Equatable {
    static func ==(lhs: ErgoBoxCandidate, rhs: ErgoBoxCandidate) -> Bool {
        ergo_wallet_ergo_box_candidate_eq(lhs.pointer, rhs.pointer)
    }
}

/// Ergo box, that is taking part in some transaction on the chain Differs with ``ErgoBoxCandidate``
/// by added transaction id and an index in the input of that transaction
class ErgoBox{
    internal var pointer: ErgoBoxPtr

    /// Make a new box.
    /// - Parameters
    ///  - `boxValue`: amount of money associated with the box
    ///  - `creationHeight`: height when a transaction containing the box is created.
    ///  - `contract`: guarding contract(``Contract``), which should be evaluated to true in order
    ///    to open(spend) this box
    ///  - `txId`: transaction id in which this box was "created" (participated in outputs)
    ///  - `index`: index (in outputs) in the transaction
    init( boxValue: BoxValue,
          creationHeight: UInt32,
          contract: Contract,
          txId: TxId,
          index: UInt16,
          tokens: Tokens
    ) throws {
        var ptr: ErgoBoxPtr?
        let error = ergo_wallet_ergo_box_new(
            boxValue.pointer,
            creationHeight,
            contract.pointer,
            txId.pointer,
            index,
            tokens.pointer,
            &ptr)
        try checkError(error)
        self.pointer = ptr!
    }
    
    /// Parse from JSON.  Supports Ergo Node/Explorer API and box values and token amount encoded as
    /// strings.
    init(withJson json: String) throws {
        var ptr: ErgoBoxPtr?
        let error = json.withCString { cs in
            ergo_wallet_ergo_box_from_json(cs, &ptr)
        }
        try checkError(error)
        self.pointer = ptr!
    }
    
    /// Takes ownership of an existing ``ErgoBoxPtr``. Note: we must ensure that no other instance
    /// of ``ErgoBox`` can hold this pointer.
    internal init(withRawPointer pointer: ErgoBoxPtr) {
        self.pointer = pointer
    }
    
    /// Get box id
    func getBoxId() -> BoxId {
        var ptr: BoxIdPtr?
        ergo_wallet_ergo_box_id(self.pointer, &ptr)
        return BoxId(withRawPointer: ptr!)
    }
    
    /// Get box creation height
    func getCreationHeight() -> UInt32 {
        return ergo_wallet_ergo_box_creation_height(self.pointer)
    }
    
    /// Get tokens for box
    func getTokens() -> Tokens {
        var tokensPtr: TokensPtr?
        ergo_wallet_ergo_box_tokens(self.pointer, &tokensPtr)
        return Tokens(withPtr: tokensPtr!)
    }
    
    /// Get ergo tree for box
    func getErgoTree() -> ErgoTree {
        var ergoTreePtr: ErgoTreePtr?
        ergo_wallet_ergo_box_ergo_tree(self.pointer, &ergoTreePtr)
        return ErgoTree(withRawPointer: ergoTreePtr!)
    }
    
    /// Get box value
    func getBoxValue() -> BoxValue {
        var boxValuePtr: BoxValuePtr?
        ergo_wallet_ergo_box_value(self.pointer, &boxValuePtr)
        return BoxValue(withRawPointer: boxValuePtr!)
    }
    
    /// Returns value (ErgoTree constant) stored in the register or `nil` if the register is empty
    func getRegisterValue(registerId: NonMandatoryRegisterId) -> Constant? {
        var constantPtr: ConstantPtr?
        let res = ergo_wallet_ergo_box_register_value(self.pointer, registerId.rawValue, &constantPtr)
        assert(res.error == nil)
        if res.is_some {
            return Constant(withRawPointer: constantPtr!)
        } else {
            return nil
        }
    }
    
    /// JSON representation as text (compatible with Ergo Node/Explorer API, numbers are encoded as numbers)
    func toJSON() throws -> JSON? {
        var cStr: UnsafePointer<CChar>?
        let error = ergo_wallet_ergo_box_to_json(self.pointer, &cStr)
        try checkError(error)
        let str = String(cString: cStr!)
        ergo_wallet_delete_string(UnsafeMutablePointer(mutating: cStr))
        return try str.data(using: .utf8, allowLossyConversion: false).map {
            try JSON(data: $0)
        }
    }
    
    /// JSON representation according to EIP-12 <https://github.com/ergoplatform/eips/pull/23>
    func toJsonEIP12() throws -> JSON? {
        var cStr: UnsafePointer<CChar>?
        let error = ergo_wallet_ergo_box_to_json_eip12(self.pointer, &cStr)
        try checkError(error)
        let str = String(cString: cStr!)
        ergo_wallet_delete_string(UnsafeMutablePointer(mutating: cStr))
        return try str.data(using: .utf8, allowLossyConversion: false).map {
            try JSON(data: $0)
        }
    }
    
    deinit {
        ergo_wallet_ergo_box_delete(self.pointer)
    }
}

extension ErgoBox: Equatable {
    static func ==(lhs: ErgoBox, rhs: ErgoBox) -> Bool {
        ergo_wallet_ergo_box_eq(lhs.pointer, rhs.pointer)
    }
}

/// Pair of (value, tokens) for a box
class ErgoBoxAssetsData {
    internal var pointer: ErgoBoxAssetsDataPtr

    /// Create new instance
    init( boxValue: BoxValue,
          tokens: Tokens
    ) {
        var ptr: ErgoBoxAssetsDataPtr?
        ergo_wallet_ergo_box_assets_data_new(
            boxValue.pointer,
            tokens.pointer,
            &ptr)
        self.pointer = ptr!
    }
    
    /// Takes ownership of an existing ``ErgoBoxAssetsDataPtr``. Note: we must ensure that no other
    /// instance of ``ErgoBoxAssetsData`` can hold this pointer.
    internal init(withRawPointer pointer: ErgoBoxAssetsDataPtr) {
        self.pointer = pointer
    }
    
    /// Value part of the box
    func getBoxValue() -> BoxValue {
        var boxValuePtr: BoxValuePtr?
        ergo_wallet_ergo_box_assets_data_value(self.pointer, &boxValuePtr)
        return BoxValue(withRawPointer: boxValuePtr!)
    }
    
    /// Tokens part of the box
    func getTokens() -> Tokens {
        var tokensPtr: TokensPtr?
        ergo_wallet_ergo_box_assets_data_tokens(self.pointer, &tokensPtr)
        return Tokens(withPtr: tokensPtr!)
    }
    
    deinit {
        ergo_wallet_ergo_box_assets_data_delete(self.pointer)
    }
}

extension ErgoBoxAssetsData: Equatable {
    static func ==(lhs: ErgoBoxAssetsData, rhs: ErgoBoxAssetsData) -> Bool {
        ergo_wallet_ergo_box_assets_data_eq(lhs.pointer, rhs.pointer)
    }
}

/// An ordered collection of ``ErgoBoxAssetsData``s
class ErgoBoxAssetsDataList {
    internal var pointer: ErgoBoxAssetsDataListPtr
    
    /// Create an empty collection
    init() {
        var ptr: ErgoBoxAssetsDataListPtr?
        ergo_wallet_ergo_box_assets_data_list_new(&ptr)
        self.pointer = ptr!
    }
    
    /// Takes ownership of an existing ``ErgoBoxAssetsDataListPtr``. Note: we must ensure that no other
    /// instance of ``ErgoBoxAssetsDataList`` can hold this pointer.
    init(withRawPointer ptr: ErgoBoxAssetsDataListPtr) {
        self.pointer = ptr
    }
    
    /// Return the length of the collection
    func len() -> UInt {
        return ergo_wallet_ergo_box_assets_data_list_len(self.pointer)
    }
    
    /// Returns the ``ErgoBoxAssetsData`` at location `index` if it exists.
    func get(index: UInt) -> ErgoBoxAssetsData? {
        var ptr: ErgoBoxAssetsDataPtr?
        let res = ergo_wallet_ergo_box_assets_data_list_get(self.pointer, index, &ptr)
        assert(res.error == nil)
        if res.is_some {
            return ErgoBoxAssetsData(withRawPointer: ptr!)
        } else {
            return nil
        }
    }
    
    /// Add a ``ErgoBoxAssetsData`` to the end of the collection.
    func add(ergoBoxAssetData: ErgoBoxAssetsData) {
        ergo_wallet_ergo_box_assets_data_list_add(ergoBoxAssetData.pointer, self.pointer)
    }
        
    deinit {
        ergo_wallet_ergo_box_assets_data_list_delete(self.pointer)
    }
}

/// Type for representing box registers R4 - R9
enum NonMandatoryRegisterId: UInt8 {
    /// id for R4 register
    case R4 = 4
    /// id for R5 register
    case R5 = 5
    /// id for R6 register
    case R6 = 6
    /// id for R7 register
    case R7 = 7
    /// id for R8 register
    case R8 = 8
    /// id for R9 register
    case R9 = 9
}

/// An ordered collection of ``ErgoBoxCandidate``s
class ErgoBoxCandidates {
    internal var pointer: ErgoBoxCandidatesPtr
    
    /// Create an empty collection
    init() {
        var ptr: ErgoBoxCandidatesPtr?
        ergo_wallet_ergo_box_candidates_new(&ptr)
        self.pointer = ptr!
    }
    
    /// Takes ownership of an existing ``ErgoBoxCandidatesPtr``. Note: we must ensure that no other
    /// instance of ``ErgoBoxCandidates`` can hold this pointer.
    init(withRawPointer ptr: ErgoBoxCandidatesPtr) {
        self.pointer = ptr
    }
    
    /// Return the length of the collection
    func len() -> UInt {
        return ergo_wallet_ergo_box_candidates_len(self.pointer)
    }
    
    /// Returns the ``ErgoBoxCandidates`` at location `index` if it exists.
    func get(index: UInt) -> ErgoBoxCandidate? {
        var dataInputPtr: DataInputPtr?
        let res = ergo_wallet_ergo_box_candidates_get(self.pointer, index, &dataInputPtr)
        assert(res.error == nil)
        if res.is_some {
            return ErgoBoxCandidate(withRawPointer: dataInputPtr!)
        } else {
            return nil
        }
    }
    
    /// Add a ``ErgoBoxCandidates`` to the end of the collection.
    func add(ergoBoxCandidate: ErgoBoxCandidate) {
        ergo_wallet_ergo_box_candidates_add(ergoBoxCandidate.pointer, self.pointer)
    }
        
    deinit {
        ergo_wallet_ergo_box_candidates_delete(self.pointer)
    }
}

/// An ordered collection of ``ErgoBox``s
class ErgoBoxes {
    internal var pointer: ErgoBoxesPtr
    
    init() {
        self.pointer = ErgoBoxes.initRawPtrEmpty()
    }
    
    /// Parse ``ErgoBox`` array from JSON (Node API)
    init(fromJSON: Any) throws {
        let json = JSON(fromJSON)
        if let arr = json.array {
            let boxes = try arr.map{ elem -> ErgoBox in
                if let jsonStr = elem.rawString() {
                    return try ErgoBox(withJson: jsonStr);
                } else {
                    throw WalletError.walletCError(reason: "Ergoboxes.init(fromJSON): cannot cast array element to raw JSON string")
                }
            }
            self.pointer = ErgoBoxes.initRawPtrEmpty()
            for ergoBox in boxes {
                self.add(ergoBox: ergoBox)
            }
        } else {
            throw WalletError.walletCError(reason: "Ergoboxes.init(fromJSON): expected [JSON]")
        }
    }
    
    /// Takes ownership of an existing ``ErgoBoxesPtr``. Note: we must ensure that no other
    /// instance of ``ErgoBoxes`` can hold this pointer.
    init(withRawPointer ptr: ErgoBoxesPtr) {
        self.pointer = ptr
    }
    
    /// Use the C-API to create an empty collection and return the raw pointer that points to this
    /// collection.
    private static func initRawPtrEmpty() -> ErgoBoxesPtr {
        var ptr: ErgoBoxesPtr?
        ergo_wallet_ergo_boxes_new(&ptr)
        return ptr!
    }
    
    /// Return the length of the collection
    func len() -> UInt {
        return ergo_wallet_ergo_boxes_len(self.pointer)
    }
    
    /// Returns the ``BlockHeader`` at location `index` if it exists.
    func get(index: UInt) -> ErgoBox? {
        var dataInputPtr: DataInputPtr?
        let res = ergo_wallet_ergo_boxes_get(self.pointer, index, &dataInputPtr)
        assert(res.error == nil)
        if res.is_some {
            return ErgoBox(withRawPointer: dataInputPtr!)
        } else {
            return nil
        }
    }
    
    /// Add a ``ErgoBox`` to the end of the collection.
    func add(ergoBox: ErgoBox) {
        ergo_wallet_ergo_boxes_add(ergoBox.pointer, self.pointer)
    }
        
    deinit {
        ergo_wallet_ergo_boxes_delete(self.pointer)
    }
}
